ENH: __pandas_priority__ (#48347)
pandas-dev/pandas: doc/source/development/extending.rst

More information on how to implement a third-party plotting backend can be found at
https://github.com/pandas-dev/pandas/blob/main/pandas/plotting/__init__.py#L1.

.. _extending.pandas_priority:

Arithmetic with 3rd party types
-------------------------------

In order to control how arithmetic works between a custom type and a pandas type,
implement ``__pandas_priority__``.  Similar to numpy's ``__array_priority__``
semantics, arithmetic methods on :class:`DataFrame`, :class:`Series`, and :class:`Index`

> objects will return ``NotImplemented`` if the ``other`` has a higher ``__array_priority__``.
> 
> .. code-block:: python
> 
>     class MyClass:
>         __pandas_priority__ = 5000
> 
>         def __add__(self, other):
>             return self
> 
>     left = MyClass()
>     right = pd.Series([1, 2, 3])
>     assert right.__add__(left) is NotImplemented
>     assert right + left is left

Reviewer: Thanks @jbrockmendel, this is very helpful.

There is a typo, `__add__` instead of `__radd__` in `MyClass`.

I also propose adding a bit more context, and some minor things that at least to me, would make this easier and faster to understand. Feel free to update as you find convenient, maybe my version it's too obvious or too verbose, but I personally thing it should help readers.

```suggestion
objects will delegate to ``other``, if it has an attribute ``__array_priority__`` with a higher value.

By default, pandas objects try to operate with other objects, even if they are not types known to pandas:

.. code-block:: python

    >>> pandas.Series([1, 2]) + [10, 20]
    0    11
    1    22
    dtype: int64

In the example above, if `[10, 20]` was a custom type that can be understood as a list, pandas objects will still operate with it in the same way.

In some cases, it is useful to delegate to the other type the operation. For example, consider I implement a
custom list object, and I want the result of adding my custom list with a pandas `Series` to be an instance of my list
and not a `Series` as seen in the previous example. This is now possible by defining the `__pandas_priority__` attribute
of my custom list, and setting it to a higher value, than the priority of the pandas objects I want to operate with.

.. code-block:: python

    class CustomList(list):
        __pandas_priority__ = 5000

        def __radd__(self, other):
            # return `self` and not the addition for simplicity
            return self

    custom = CustomList()
    series = pd.Series([1, 2, 3])

    # Series refuses to add custom, since it's an unknown type with higher priority
    assert series.__add__(custom) is NotImplemented

    # This will cause the custom class `__radd__` being used instead
    assert series + custom is custom
```

More information on how to implement a third-party plotting backend can be found at
https://github.com/pandas-dev/pandas/blob/main/pandas/plotting/__init__.py#L1.

.. _extending.pandas_priority:

Arithmetic with 3rd party types
-------------------------------

In order to control how arithmetic works between a custom type and a pandas type,
implement ``__pandas_priority__``.  Similar to numpy's ``__array_priority__``
semantics, arithmetic methods on :class:`DataFrame`, :class:`Series`, and :class:`Index`

> objects will delegate to ``other``, if it has an attribute ``__array_priority__`` with a higher value.

Reviewer 2: ```suggestion
objects will delegate to ``other``, if it has an attribute ``__pandas_priority__`` with a higher value.
```
?

Author: good catch, will update

By default, pandas objects try to operate with other objects, even if they are not types known to pandas:

.. code-block:: python

    >>> pandas.Series([1, 2]) + [10, 20]
    0    11
    1    22
    dtype: int64

> In the example above, if `[10, 20]` was a custom type that can be understood as a list, pandas objects will still operate with it in the same way.

Reviewer 2: ```suggestion
In the example above, if ``[10, 20]`` was a custom type that can be understood as a list, pandas objects will still operate with it in the same way.
```

In some cases, it is useful to delegate to the other type the operation. For example, consider I implement a

> custom list object, and I want the result of adding my custom list with a pandas `Series` to be an instance of my list

Reviewer 2: ```suggestion
custom list object, and I want the result of adding my custom list with a pandas class:`Series` to be an instance of my list
```

> and not a `Series` as seen in the previous example. This is now possible by defining the `__pandas_priority__` attribute

Reviewer 2: ```suggestion
and not a :class:`Series` as seen in the previous example. This is now possible by defining the ``__pandas_priority__`` attribute
```

> of my custom list, and setting it to a higher value, than the priority of the pandas objects I want to operate with.

Reviewer 2: Might be good to list the priority values of pandas objects here?

Author: Sure, will update
